### PR TITLE
add devops project & members & pipeline api

### PR DESCRIPTION
**Pull request by runzexia**
@zryfish @wansir

### CODE REVIEW

File: pkg/apis/devops/v1alpha2/register.go [hunk 1]
```diff
@@ -35,21 +37,71 @@ var (
...
+		Doc("get devops project").
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Param(webservice.PathParameter("devops", "projectId")).
+		Writes(&devops.DevOpsProject{}))
+
+	webservice.Route(webservice.GET("/devops/{devops}/defaultroles").
+		To(devopsapi.GetDevOpsProjectDefaultRoles).
+		Doc("get devops project defaultroles").
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Writes(&devops.DefaultRoles))
```
Comment (zryfish) [inline pkg/apis/devops/v1alpha2/register.go]: document every parameter
`Param(webservice.PathParameter("devops", projectID")`

File: pkg/apis/devops/v1alpha2/register.go [hunk 2]
```diff
@@ -35,21 +37,71 @@ var (
...
+
+	webservice.Route(webservice.GET("/devops/{devops}/members").
+		To(devopsapi.GetDevOpsProjectMembersHandler).
+		Doc("get devops project members").
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Writes(&[]*devops.DevOpsProjectMembership{}))
+
+	webservice.Route(webservice.GET("/devops/{devops}/members/{members}").
```
Comment (zryfish) [inline pkg/apis/devops/v1alpha2/register.go]: normally, parameter name use singular, like `member`

Comment (runzexia) [reply pkg/apis/devops/v1alpha2/register.go]: In k8s, the plural of the resource is always used in the path.

File: pkg/models/devops/project_member_handler.go [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+	err := dbconn.Select(DevOpsProjectMembershipColumns...).
+		From(DevOpsProjectMembershipTableName).
+		Where(db.And(
+			db.Eq(DevOpsProjectMembershipUsernameColumn, member.Username),
+			db.Eq(DevOpsProjectMembershipProjectIdColumn, projectId))).LoadOne(membership)
+	if err != nil && err != db.ErrNotFound {
+		glog.Errorf("%+v", err)
+		return nil, restful.NewError(http.StatusInternalServerError, err.Error())
+	}
+	if err != db.ErrNotFound {
```
Comment (zryfish) [inline pkg/models/devops/project_member_handler.go]: is this right? if err == nil  , will satisfy this condition and return?

File: pkg/models/devops/project_member_handler.go [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+	if err != nil {
+		glog.Errorf("%+v", err)
+		return nil, restful.NewError(utils.GetJenkinsStatusCode(err), err.Error())
+	}
+	projectMembership := NewDevOpsProjectMemberShip(member.Username, projectId, member.Role, operator)
+	_, err = dbconn.
+		InsertInto(DevOpsProjectMembershipTableName).
+		Columns(DevOpsProjectMembershipColumns...).
+		Record(projectMembership).Exec()
+	if err != nil {
```
Comment (zryfish) [inline pkg/models/devops/project_member_handler.go]: if db insertion failed, do we need to undo previous role assigning?

File: pkg/models/devops/project_member_handler.go [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+		}
+	default:
+		if reverse {
+			query.OrderDesc(DevOpsProjectMembershipRoleColumn)
+		} else {
+			query.OrderAsc(DevOpsProjectMembershipRoleColumn)
+		}
+	}
+	query.Limit(uint64(limit))
+	query.Offset(uint64(offset))
```
[the remark this answers is not in the document]
Comment (runzexia) [reply pkg/models/devops/project_member_handler.go]: in pkg/apiserver/devops/member.go Ttee parameters are parsed and  no case of <0.

File: pkg/apis/devops/v1alpha2/register.go [hunk 6]
```diff
@@ -35,21 +39,199 @@ var (
...
 
-	tags := []string{"devops"}
+	tags := []string{"DevOps"}
+
+	webservice.Route(webservice.GET("/devops/{devops}").
+		To(devopsapi.GetDevOpsProjectHandler).
+		Doc("get devops project").
+		Metadata(restfulspec.KeyOpenAPITags, tags).
+		Param(webservice.PathParameter("devops", "devops project's Id")).
+		Returns(http.StatusOK, "success", devops.DevOpsProject{}).
```
Comment (soulseen) [inline pkg/apis/devops/v1alpha2/register.go]: use `success` as constant here.  In order to keep unify,  the description message suggest use `ok`, default is `ok`.

**Pull request merged**